### DISCUSSION TITLE
Document pattern for initializing state from props

### DISCUSSION THREAD

**Discussion by alexgagnon** [lit/lit#2358] [Q&A]
Is there a preferred pattern for initializing state from properties that don't have a default value, and that without a value would throw an error in render? Here's a small snippet of what I'm running into

```ts
import {html, LitElement} from 'lit';
import {property, state} from 'lit/decorators.js'

export class Example extends LitElement {
  @property({type: Number})
  pageSize: number = 10;
  
  @property({type: Number})
  total!: number;

  @state()
  _count = Math.ceil(this.total / this.pageSize); // this is NaN on first render...

  render() {
    console.log(this._count);
    return html`
      <ul>
        ${Array.from(Array(this._count).keys()).map((_, index) => html`<li>${index}</li>`)}
      </ul>
    `;
  }
}
```

I guess intuitively I would expect a render method that depends on state to 'wait' for that state's dependencies to have been assigned (especially since this.total has no default value but is defined as being required), but from what I understand an initial first pass render will happen and that's where the issue happens. Is the best-practice here to assigned a dummy default value, or to check in render if a value is null and return null in that case? I guess I'm just looking for some docs on this pattern... thanks!

**Comment by e111077** [comment]
There is no preferred way to do this, but a good place for reactive property reconciliation to live is in `willUpdate`. There is where you should modify other reactive properties based on the changes of others.

Additionally, `shouldUpdate` is another place to determine whether an update should happen or not, but note that it is not called on the Server at SSR time unlike `willUpdate`.

[See docs for Reactive Update lifecycle](https://lit.dev/docs/components/lifecycle/#reactive-update-cycle)

Personally, if I were to tackle this I would have a check in `shouldUpdate` to return `false` if `total` is not defined but log a warning in `willUpdate` and set `total` to an initial value that will not crash in case of SSR.

**Reply by alexgagnon** [reply]
Thanks! I did a little modification which I believe should work based on your explanation. However it only seems to execute once if `shouldUpdate` returns false? Clearly I'm not doing something right :P

```ts
import {html, css, LitElement} from 'lit';
import {customElement, property, state} from 'lit/decorators.js';

@customElement('simple-greeting')
export class Example extends LitElement {
  @property({type: Number})
  pageSize: number = 10;
  
  @property({type: Number})
  total!: number;

  @state()
  _count = Math.ceil(this.total / this.pageSize); // this is NaN on first render...

  //shouldUpdate(changedProperties: Map<string, any>) {
  //  return changedProperties.get('total') != null ;
  //}

  willUpdate(changedProperties: Map<string, any>) {
    if (changedProperties.get('total') != null || changedProperties.has('pageSize')) {
      this._count = Math.ceil(this.total / this.pageSize);
    }
  }
  
  render() {
    console.log(this._count);
    return html`
      <ul>
        ${Array.from(Array(this._count).keys()).map((_, index) => html`<li>${index}</li>`)}
      </ul>
    `;
  }
}
```

**Comment by justinfagnani** [comment]
In cases where the state is a simple enough computation over properties, I think you want a getter. but you also very much should take into account when properties might not have been set from the outside yet. I would be very wary of just placing non-null-assertions (`!`) on properties - that's usually lying to the type system and not letting it tell you when you need to consider `undefined`.

```ts
import {html, LitElement} from 'lit';
import {property, state} from 'lit/decorators.js'

export class Example extends LitElement {
  @property({type: Number})
  pageSize: number = 10;
  
  @property({type: Number})
  total?: number;

  get _count() {
    return this.total === undefined || this.pageSize ? 0 : Math.ceil(this.total / this.pageSize);
  }

  render() {
    console.log(this._count);
    return html`
      <ul>
        ${Array.from(Array(this._count).keys()).map((_, index) => html`<li>${index}</li>`)}
      </ul>
    `;
  }
}
```

We don't yet have a comprehensive section for this kind of advice. It's a lot of general JS advice, which we don't always have in the docs because it's not Lit=specific. We have to figure out how to organize it. cc @arthurevans

**Reply by alexgagnon** [reply]
Thanks @justinfagnani! I think the 'lying to the type system' is what I'm getting tripped up on... In my mind since I'm writing a component like `<some-example pageSize="10" total="19"></some-example>`, and I'm saying that `this.total` is non-null, it _appears_ (to me, at least), as though I am initializing the component with `this.total` on the first pass (i.e. before first render). I'm learning that even these 'initial' props trigger an update. Is it prudent then to just assign defaults for all properties?

**Reply by blikblum** [reply]
> Thanks @justinfagnani! I think the 'lying to the type system' is what I'm getting tripped up on... In my mind since I'm writing a component like `<some-example pageSize="10" total="19"></some-example>`, and I'm saying that `this.total` is non-null, it _appears_ (to me, at least), as though I am initializing the component with `this.total` on the first pass (i.e. before first render). I'm learning that even these 'initial' props trigger an update. Is it prudent then to just assign defaults for all properties?

In the first pass, the check changedProperties.get('total') is always undefined. You need to check for total value in connected callback also.

AFAIK there's no single place where you can check for property value changes and initialization.

This is how i would implement:

Notes:
 - i removed the state decorator from _count which is unnecessary
 - changedProperties.get('total') returns the old value so is not correct
 - you may need more checks for bogus values or when total is not defined beforehand

```js
import {html, css, LitElement} from 'lit';
import {customElement, property, state} from 'lit/decorators.js';

@customElement('simple-greeting')
export class Example extends LitElement {
  @property({type: Number})
  pageSize: number = 10;
  
  @property({type: Number})
  total!: number;

  _count
  
  updateCount() {
    this._count = Math.ceil(this.total / this.pageSize);
  }
 
  connectedCallback() {
    super.connectedCallback()
    if (!this.hasUpdated && this.total !== undefined) {
       this.updateCount()
    }
  }
  

  willUpdate(changedProperties: Map<string, any>) {
    if (changedProperties.has('total') || changedProperties.has('pageSize')) {
      // this is not called when total is initially set as an attribute
      this.updateCount()
    }
  }
  
  render() {
    console.log(this._count);
    return html`
      <ul>
        ${Array.from(Array(this._count).keys()).map((_, index) => html`<li>${index}</li>`)}
      </ul>
    `;
  }
}
```

**Reply by justinfagnani** [reply]
@alexgagnon It is not necessarily true that `total` will be unset on the first render. If the user of the element set `total` when the element is created, either through a system like lit-html bindings, or in code like:

```ts
const el = new Example();
el.total = 3;
```

Then `total` _will_ be defined by the first render.

The initializer do technically cause a render, because the setters call `requestUpdate()`, but they don't usually cause any _additional_ renders because we schedule a render on connection too. You can assign initial values to properties, and that's often really good to do.

**Reply by justinfagnani** [reply]
@blikblum for something like `_count` here, I think it's overkill to programmatically update it. Computing on the fly is less code and likely faster.

**Reply by blikblum** [reply]
Yes. But the question still stand for more computationally heavy situations. I use this pattern for cases like transforming arrays where creating new arrays on the fly is not desired.

Actual code:

```js
class SectorsInput extends CustomArrayInput {
  @property({ attribute: false })
  groupItems

  _groupItems
  
  connectedCallback() {
    super.connectedCallback()
    this.normalizeGroupItems()
  }

  normalizeGroupItems() {
    this._groupItems = this.groupItems
      ? this.groupItems.map((item) => ({ name: item.name, value: item.id }))
      : []
  }

  // using lit-element 2.x . no willUpdate
  update(changed) {
    if (changed.has('groupItems')) {
      this.normalizeGroupItems()
    }
    super.update(changed)
  }


  render () {
    return html`<div class="card mb-0">
              <checkbox-group-input
                name="groups"
                custom-input
                .items=${this._groupItems}                
              ></checkbox-group-input>           
    </div>`
  } 
}
```

**Reply by alexgagnon** [reply]
Thanks everyone, so to summarize:

- This will not have a value on first render:
`<some-example pageSize="10" total="19"></some-example>`

- These will have it:
``html`<some-example .pageSize="10" .total="19"></some-example>` `` (either property or attribute should work)
```js
const example = new Example();
example.total = 19;
```

- Use a getter for simple computed properties

- For more involved cases, you can initialize/update using the `connectedCallback` and `willUpdate` (both of which will still work with SSR)

- Default values for properties can be a safe way to avoid this behaviour

Is that along the right track?

**Reply by blikblum** [reply]
> * This will not have a value on first render:
>   `<some-example pageSize="10" total="19"></some-example>`
> 
> * These will have it:
>   `` html`<some-example .pageSize="10" .total="19"></some-example>`  ``- (using properties, attributes will not have it on first render)

AFAIK, setting through property or attribute should exhibit the same behavior

**Reply by alexgagnon** [reply]
updated

**Reply by justinfagnani** [reply]
@alexgagnon Attributes from HTML will be available first render too. I'm confused why you're thinking that externally set values won't be available on the first render. Can you make an example on the playground to show what you're experiencing?

**Reply by alexgagnon** [reply]
Ah my mistake @justinfagnani , I assumed based on your comment:

> If the user of the element set total when the element is created, either through a system like lit-html bindings, or in code like [...]. Then total will be defined by the first render.

and the one from @blikblum's code in `willUpdate`:

>       // this is not called when total is initially set as an attribute

that if we did not use lit-html bindings or imperatively set the property (as in, not through plain HTML attributes), that it wouldn't work. I created this playground though that shows both working [here](https://lit.dev/playground/#project=W3sibmFtZSI6IkJhc2ljLnRzIiwiY29udGVudCI6ImltcG9ydCB7aHRtbCwgY3NzLCBMaXRFbGVtZW50fSBmcm9tICdsaXQnO1xuaW1wb3J0IHtjdXN0b21FbGVtZW50LCBwcm9wZXJ0eSwgc3RhdGV9IGZyb20gJ2xpdC9kZWNvcmF0b3JzLmpzJztcblxuQGN1c3RvbUVsZW1lbnQoJ2Jhc2ljLWV4YW1wbGUnKVxuZXhwb3J0IGNsYXNzIEJhc2ljRXhhbXBsZSBleHRlbmRzIExpdEVsZW1lbnQge1xuICBAcHJvcGVydHkoe3R5cGU6IE51bWJlcn0pXG4gIHBhZ2VTaXplOiBudW1iZXIgPSAxMDtcbiAgXG4gIEBwcm9wZXJ0eSh7dHlwZTogTnVtYmVyfSlcbiAgdG90YWwhOiBudW1iZXI7XG5cbiAgQHN0YXRlKClcbiAgX2NvdW50ID0gTWF0aC5jZWlsKHRoaXMudG90YWwgLyB0aGlzLnBhZ2VTaXplKTtcblxuICByZW5kZXIoKSB7XG4gICAgY29uc29sZS5sb2codGhpcy5fY291bnQpO1xuICAgIHJldHVybiBodG1sYFxuICAgICAgPHVsPlxuICAgICAgICAke0FycmF5LmZyb20oQXJyYXkodGhpcy5fY291bnQpLmtleXMoKSkubWFwKChfLCBpbmRleCkgPT4gaHRtbGA8bGk-JHtpbmRleH08L2xpPmApfVxuICAgICAgPC91bD5cbiAgICBgO1xuICB9XG59In0seyJuYW1lIjoiaW5kZXguaHRtbCIsImNvbnRlbnQiOiI8IURPQ1RZUEUgaHRtbD5cbjxoZWFkPlxuICA8c2NyaXB0IHR5cGU9XCJtb2R1bGVcIiBzcmM9XCIuL0Jhc2ljLmpzXCI-PC9zY3JpcHQ-XG4gIDxzY3JpcHQgdHlwZT1cIm1vZHVsZVwiIHNyYz1cIi4vR2V0dGVyLmpzXCI-PC9zY3JpcHQ-XG4gIDxzY3JpcHQgdHlwZT1cIm1vZHVsZVwiIHNyYz1cIi4vTGlmZWN5Y2xlLmpzXCI-PC9zY3JpcHQ-XG48L2hlYWQ-XG48Ym9keT5cbiAgPHA-QmFzaWM8L3A-XG4gIDxiYXNpYy1leGFtcGxlIHBhZ2VTaXplPVwiMTBcIiB0b3RhbD1cIjE5XCI-PC9iYXNpYy1leGFtcGxlPlxuICA8cD5HZXR0ZXI8L3A-XG4gIDxnZXR0ZXItZXhhbXBsZSBwYWdlU2l6ZT1cIjEwXCIgdG90YWw9XCIxOVwiPjwvZ2V0dGVyLWV4YW1wbGU-XG4gIDxwPkxpZmVjeWNsZTwvcD5cbiAgPGxpZmVjeWNsZS1leGFtcGxlIHBhZ2VTaXplPVwiMTBcIiB0b3RhbD1cIjE5XCI-PC9saWZlY3ljbGUtZXhhbXBsZT5cbjwvYm9keT5cbiJ9LHsibmFtZSI6InBhY2thZ2UuanNvbiIsImNvbnRlbnQiOiJ7XG4gIFwiZGVwZW5kZW5jaWVzXCI6IHtcbiAgICBcImxpdFwiOiBcIl4yLjAuMFwiLFxuICAgIFwiQGxpdC9yZWFjdGl2ZS1lbGVtZW50XCI6IFwiXjEuMC4wXCIsXG4gICAgXCJsaXQtZWxlbWVudFwiOiBcIl4zLjAuMFwiLFxuICAgIFwibGl0LWh0bWxcIjogXCJeMi4wLjBcIlxuICB9XG59IiwiaGlkZGVuIjp0cnVlfSx7Im5hbWUiOiJHZXR0ZXIudHMiLCJjb250ZW50IjoiaW1wb3J0IHtodG1sLCBjc3MsIExpdEVsZW1lbnR9IGZyb20gJ2xpdCc7XG5pbXBvcnQge2N1c3RvbUVsZW1lbnQsIHByb3BlcnR5LCBzdGF0ZX0gZnJvbSAnbGl0L2RlY29yYXRvcnMuanMnO1xuXG5AY3VzdG9tRWxlbWVudCgnZ2V0dGVyLWV4YW1wbGUnKVxuZXhwb3J0IGNsYXNzIEdldHRlckV4YW1wbGUgZXh0ZW5kcyBMaXRFbGVtZW50IHtcbiAgQHByb3BlcnR5KHt0eXBlOiBOdW1iZXJ9KVxuICBwYWdlU2l6ZTogbnVtYmVyID0gMTA7XG4gIFxuICBAcHJvcGVydHkoe3R5cGU6IE51bWJlcn0pXG4gIHRvdGFsITogbnVtYmVyO1xuXG4gIGdldCBfY291bnQoKSB7XG4gICAgcmV0dXJuIHRoaXMudG90YWwgPT0gbnVsbCA_IDAgOiBNYXRoLmNlaWwodGhpcy50b3RhbCAvIHRoaXMucGFnZVNpemUpO1xuICB9XG4gICAgXG4gIHJlbmRlcigpIHtcbiAgICBjb25zb2xlLmxvZyh0aGlzLl9jb3VudCk7XG4gICAgcmV0dXJuIGh0bWxgXG4gICAgICA8dWw-XG4gICAgICAgICR7QXJyYXkuZnJvbShBcnJheSh0aGlzLl9jb3VudCkua2V5cygpKS5tYXAoKF8sIGluZGV4KSA9PiBodG1sYDxsaT4ke2luZGV4fTwvbGk-YCl9XG4gICAgICA8L3VsPlxuICAgIGA7XG4gIH1cbn0ifSx7Im5hbWUiOiJMaWZlY3ljbGUudHMiLCJjb250ZW50IjoiaW1wb3J0IHtodG1sLCBjc3MsIExpdEVsZW1lbnR9IGZyb20gJ2xpdCc7XG5pbXBvcnQge2N1c3RvbUVsZW1lbnQsIHByb3BlcnR5LCBzdGF0ZX0gZnJvbSAnbGl0L2RlY29yYXRvcnMuanMnO1xuXG5AY3VzdG9tRWxlbWVudCgnbGlmZWN5Y2xlLWV4YW1wbGUnKVxuZXhwb3J0IGNsYXNzIExpZmVjeWNsZUV4YW1wbGUgZXh0ZW5kcyBMaXRFbGVtZW50IHtcbiAgQHByb3BlcnR5KHt0eXBlOiBOdW1iZXJ9KVxuICBwYWdlU2l6ZTogbnVtYmVyID0gMTA7XG4gIFxuICBAcHJvcGVydHkoe3R5cGU6IE51bWJlcn0pXG4gIHRvdGFsITogbnVtYmVyO1xuXG4gIF9jb3VudDtcbiAgXG4gIHVwZGF0ZUNvdW50KCkge1xuICAgIHRoaXMuX2NvdW50ID0gTWF0aC5jZWlsKHRoaXMudG90YWwgLyB0aGlzLnBhZ2VTaXplKTtcbiAgfVxuLyogXG4gIGNvbm5lY3RlZENhbGxiYWNrKCkge1xuICAgIHN1cGVyLmNvbm5lY3RlZENhbGxiYWNrKClcbiAgICBpZiAoIXRoaXMuaGFzVXBkYXRlZCAmJiB0aGlzLnRvdGFsICE9PSB1bmRlZmluZWQpIHtcbiAgICAgICB0aGlzLnVwZGF0ZUNvdW50KClcbiAgICB9XG4gIH1cbiovXG4gIHdpbGxVcGRhdGUoY2hhbmdlZFByb3BlcnRpZXM6IE1hcDxzdHJpbmcsIGFueT4pIHtcbiAgICBpZiAoY2hhbmdlZFByb3BlcnRpZXMuaGFzKCd0b3RhbCcpIHx8IGNoYW5nZWRQcm9wZXJ0aWVzLmhhcygncGFnZVNpemUnKSkge1xuICAgICAgLy8gdGhpcyBpcyBub3QgY2FsbGVkIHdoZW4gdG90YWwgaXMgaW5pdGlhbGx5IHNldCBhcyBhbiBhdHRyaWJ1dGVcbiAgICAgIHRoaXMudXBkYXRlQ291bnQoKVxuICAgIH1cbiAgfVxuICBcbiAgcmVuZGVyKCkge1xuICAgIGNvbnNvbGUubG9nKHRoaXMuX2NvdW50KTtcbiAgICByZXR1cm4gaHRtbGBcbiAgICAgIDx1bD5cbiAgICAgICAgJHtBcnJheS5mcm9tKEFycmF5KHRoaXMuX2NvdW50KS5rZXlzKCkpLm1hcCgoXywgaW5kZXgpID0-IGh0bWxgPGxpPiR7aW5kZXh9PC9saT5gKX1cbiAgICAgIDwvdWw-XG4gICAgYDtcbiAgfVxufSJ9XQ)

**Reply by blikblum** [reply]
> // this is not called when total is initially set as an attribute

This seems not true with lit 2.0.

Played a bit and in first render, the changed map contains the assigned property even when setting via attribute, so no need to do the check at connectedCallback. Sorry about the noise

https://lit.dev/playground/#project=W3sibmFtZSI6InNpbXBsZS1ncmVldGluZy50cyIsImNvbnRlbnQiOiJpbXBvcnQge2h0bWwsIGNzcywgTGl0RWxlbWVudH0gZnJvbSAnbGl0JztcbmltcG9ydCB7Y3VzdG9tRWxlbWVudCwgcHJvcGVydHl9IGZyb20gJ2xpdC9kZWNvcmF0b3JzLmpzJztcblxuQGN1c3RvbUVsZW1lbnQoJ3NpbXBsZS1ncmVldGluZycpXG5leHBvcnQgY2xhc3MgU2ltcGxlR3JlZXRpbmcgZXh0ZW5kcyBMaXRFbGVtZW50IHtcbiAgc3RhdGljIHN0eWxlcyA9IGNzc2BwIHsgY29sb3I6IGJsdWUgfWA7XG5cbiAgQHByb3BlcnR5KClcbiAgbmFtZSBcbiAgXG4gIHdpbGxVcGRhdGUoY2hhbmdlZCkge1xuICAgICBpZiAoY2hhbmdlZC5oYXMoJ25hbWUnKSkge1xuICAgICAgIHRoaXMuZGVyaXZlZE5hbWUgPSAnZGVyaXZlZCdcbiAgICAgfVxuICB9XG5cbiAgcmVuZGVyKCkge1xuICAgIHJldHVybiBodG1sYDxwPkhlbGxvLCAke3RoaXMuZGVyaXZlZE5hbWV9ITwvcD5gO1xuICB9XG59XG4ifSx7Im5hbWUiOiJpbmRleC5odG1sIiwiY29udGVudCI6IjwhRE9DVFlQRSBodG1sPlxuPGhlYWQ-XG4gIDxzY3JpcHQgdHlwZT1cIm1vZHVsZVwiIHNyYz1cIi4vc2ltcGxlLWdyZWV0aW5nLmpzXCI-PC9zY3JpcHQ-XG48L2hlYWQ-XG48Ym9keT5cbiAgPHNpbXBsZS1ncmVldGluZyBuYW1lPVwiV29ybGRcIj48L3NpbXBsZS1ncmVldGluZz5cbjwvYm9keT5cbiJ9LHsibmFtZSI6InBhY2thZ2UuanNvbiIsImNvbnRlbnQiOiJ7XG4gIFwiZGVwZW5kZW5jaWVzXCI6IHtcbiAgICBcImxpdFwiOiBcIl4yLjAuMFwiLFxuICAgIFwiQGxpdC9yZWFjdGl2ZS1lbGVtZW50XCI6IFwiXjEuMC4wXCIsXG4gICAgXCJsaXQtZWxlbWVudFwiOiBcIl4zLjAuMFwiLFxuICAgIFwibGl0LWh0bWxcIjogXCJeMi4wLjBcIlxuICB9XG59IiwiaGlkZGVuIjp0cnVlfV0
